### DISCUSSION TITLE
Failure to converge and negative jacobian when using TemperatureDependentStressUpdate

### DISCUSSION THREAD

**Discussion by Edward-Eth** [idaholab/moose#21563] [Q&A Modules: Solid mechanics] [answered]
I have been having issues with the stability of a plastic model and have made a small input to test the issue. I have found that convergence is very unreliable and that I am sometimes getting negative jacobians? I would appreciate anyone who can look over this input file to give me any insight into the issues I am having. At time step 12 I get a negative jacobian in element 124.

<details>
  <summary>Input File</summary>

```
TEMPERATURE = 200

[Mesh]
  [Block]
    type = GeneratedMeshGenerator
    dim = 3
    xmax = 10
    ymax = 10
    zmax = 10
    nx = 5
    ny = 5
    nz = 5
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[AuxVariables]
  [Temp]
    order = FIRST
    family = LAGRANGE
    initial_condition = ${TEMPERATURE}
  []
[]

[Functions]
  [Load]
    type = PiecewiseLinear
    x = '0 4    5 10'
    y = '0 1800 0 0'
  []
  [hf_0]
    type = PiecewiseLinear
    x = '0.0 0.5'
    y = '1000 1600'
    extrap = true
  []
  [hf_100]
    type = PiecewiseLinear
    x = '0.0 0.5'
    y = '800 1300'
    extrap = true
  []
  [hf_200]
    type = PiecewiseLinear
    x = '0.0 0.5'
    y = '600 1000'
    extrap = true
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    incremental = true
    add_variables = true
    use_finite_deform_jacobian = true
    # volumetric_locking_correction = true
    use_automatic_differentiation = true
  []
[]

[BCs]
  [Fixed_X]
    type = ADDirichletBC
    boundary = 'back front'
    value = 0
    variable = disp_x
  []
  [Fixed_Y]
    type = ADDirichletBC
    boundary = 'back front'
    value = 0
    variable = disp_y
  []
  [Fixed_Z]
    type = ADDirichletBC
    boundary = back
    value = 0
    variable = disp_z
  []
  [Force_BC]
    type = ADFunctionNeumannBC
    variable = disp_z
    function = Load
    boundary = front
  []
[]

[Materials]
  [Titanium_Elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 95000
    poissons_ratio = 0.33
  []
  [Titanium_Plasticity_Stress_Update]
    type = ADTemperatureDependentHardeningStressUpdate
    hardening_functions = "hf_0 hf_100 hf_200"
    temperatures = "0 100 200"
    temperature = Temp
  []
  [Titanium_Radial_Return_Stress]
    type = ADComputeMultipleInelasticStress
    inelastic_models = 'Titanium_Plasticity_Stress_Update'
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'

  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-ksp_gmres_restart -pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter'
  petsc_options_value = '201                hypre    boomeramg      10'

  line_search = 'none'

  l_max_its = 100
  nl_max_its = 100
  nl_rel_tol = 1e-12
  nl_abs_tol = 1e-10
  l_tol = 1e-9

  start_time = 0.0
  end_time = 5.5
  dt = 0.25
[]

[Outputs]
  [outfile]
    type = Exodus
  []
[]
```
</details>

**Comment by GiudGiud** [comment]
Hello

A negative Jacobian would mean an element got deformed enough to be inverted. 
Your mesh is really coarse and your time steps are pretty large. Could you please check if those parameters are the issue?

Guillaume

**Reply by Edward-Eth** [reply]
Decreasing timesteps and increasing element counts to 0.1 seconds and 15 per side respectively has the effect of changing when convergence fails, but doesn't fix the problem. This kind of behaviour is something I've been fighting with in larger models for a while.

**Reply by GiudGiud** [reply]
@jiangwen84 could you please take a look

**Comment by Edward-Eth** [comment]
Is it possible to get MOOSE to output the results of an unconverged timestep to see what has happened leading to non-convergence?

**Reply by GiudGiud** [reply]
yes but it s not very easy to do.
Save the result of a calculation right before it fails, using checkpoint for example. 

Then load the calculation and do a single time step and set the nl_abs_tol and nl_rel_tol loose enough to always accept the result

**Reply by GiudGiud** [reply]
then look at the output

**Comment by hugary1995** [comment, marked as answer]
The convergence issue here is caused by an inappropriate choice of the path following algorithm.

The load displacement curve of a plastic material typically looks like this
![image](https://user-images.githubusercontent.com/31555204/178550888-af081acb-6c8d-478e-aa2e-2281287aa7ab.png)

In your input file you are using a Neumann BC to "follow" this curve, which is equivalent to prescribing the y component of this curve. It is apparent that no solution exist pass the ultimate strength. Numerically, this relates to non-convergence with many inverted elements. You are essentially asking the model to perform an impossible task.

In this case, you'll have to change the path following algorithm to a displacement-controlled one. That is changing the Neumann BC to a Dirichlet BC. You can add a postprocessor to output the force:
```python
[AuxVariables]
  [fx]
  []
  [fy]
  []
  [fz]
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    ... # same as before
    save_in = 'fx fy fz'
  []
[]

[Postprocessors]
  [force]
    type = NodalSum
    variable = fz
    boundary = front
  []
[]
```

You'll see that the `force` starts decreasing after certain point, and that corresponds to the point at which Neumann BC fails.

**Reply by Edward-Eth** [reply]
My plasticity curves/hardening functions don't look like that though. I'm effectively using an elastic bilinear plasticity, which looks like this: 
![image](https://user-images.githubusercontent.com/64844573/178682841-6e094c54-da4a-466b-b4a2-b8c7278dc67b.png)
The hardening functions are linear and extrapolated so increasing strain will always require increasing stress. This was a simplification I made to attempt to get the plasticity to converge without any success.
EDIT: To be clear this isn't a plot of my exact functions, just an image off google. I have a youngs modulus of 95000, tangent modulus of 1200 at 0 degrees, 1000 at 100 degrees and 800 at 200 degrees, with yield stresses of 1000, 800 and 600 respectively.

**Reply by Edward-Eth** [reply]
I've just attempted to use generate_ouput in my tensor mechanics block to output effective_plastic_strain and gotten the following error:
```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

The following material properties are declared on block 0 by multiple materials:
  Material Property             Material Objects
  effective_plastic_strain      Titanium_Plasticity_Stress_Update all_strain effective_plastic_strain_all
```
Could this have something to do with my issue? If plasticity is being defined in multiple locations?

**Reply by Edward-Eth** [reply]
So my element max effective plastic strain looks like this at the time step before the one that fails to converge:
```
Postprocessor Values:
+----------------+----------------+----------------+
| time           | force          | plastic        |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e-01 |  -4.500000e+01 |   0.000000e+00 |
|   2.000000e-01 |  -9.000000e+01 |   0.000000e+00 |
|   3.000000e-01 |  -1.350000e+02 |   0.000000e+00 |
|   4.000000e-01 |  -1.800000e+02 |   0.000000e+00 |
|   5.000000e-01 |  -2.250000e+02 |   0.000000e+00 |
|   6.000000e-01 |  -2.700000e+02 |   0.000000e+00 |
|   7.000000e-01 |  -3.150000e+02 |   0.000000e+00 |
|   8.000000e-01 |  -3.600000e+02 |   0.000000e+00 |
|   9.000000e-01 |  -4.050000e+02 |   6.720581e-05 |
|   1.000000e+00 |  -4.500000e+02 |   3.137183e-04 |
|   1.100000e+00 |  -4.950000e+02 |   8.767617e-04 |
|   1.200000e+00 |  -5.400000e+02 |   1.637166e-03 |
|   1.300000e+00 |  -5.850000e+02 |   2.639385e-03 |
+----------------+----------------+----------------+
```
2.64e-3 is well within the positive slope defined by my hardening functions, they go up to 0.5 and then extrapolate so in theory climb for ever. So I don't think I should have any elements distorting massively leading to non-convergence.

**Reply by hugary1995** [reply]
You are mixing the concept of material point constitutive model and macroscopic load displacement curve. Linear hardening still gives rise to necking. Please try my suggestion above.
